tools/wake-format/emitter.cpp

```diff
@@ -0,0 +1,217 @@
+/*
+ * Copyright 2022 SiFive, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You should have received a copy of LICENSE.Apache2 along with
+ * this software. If not, you may obtain a copy at
+ *
+ *    https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// Open Group Base Specifications Issue 7
+#define _XOPEN_SOURCE 700
+#define _POSIX_C_SOURCE 200809L
+
+#include "emitter.h"
+
+#include <cassert>
+
+#include "parser/parser.h"
+
+wcl::rope Emitter::newline(ctx_t ctx) {
+  wcl::rope_builder builder;
+
+  builder.append("⏎\n");
+  for (int i = 0; i < ctx.nest_level; i++) {
+    builder.append(space(space_per_indent));
+  }
+
+  return std::move(builder).build();
+}
+
+wcl::rope Emitter::space(uint8_t count) {
+  wcl::rope_builder builder;
+  for (uint8_t i = 0; i < count; i++) {
+    builder.append("·");
+  }
+  return std::move(builder).build();
+}
+
+wcl::optional<wcl::rope> Emitter::flat(ctx_t ctx, CSTElement node) {
+  return walk_node(ctx.flat(), node);
+}
+
+wcl::rope Emitter::try_flat(ctx_t ctx, CSTElement node) {
+  auto flat_node = flat(ctx, node);
+  if (flat_node && flat_node->size() + ctx.nest_level * space_per_indent < 80) {
+    return *flat_node;
+  }
+  auto full_node = walk_node(ctx, node);
+  assert(full_node);
+  return *full_node;
+}
+
+#define ASSERT_TOKEN(node, token) \
+  assert(!node.empty());          \
+  assert(node.id() == token);
+
+wcl::rope Emitter::layout(CST cst) {
+  ctx_t ctx;
+  return walk_top(ctx, cst.root()).concat(newline(ctx));
+}
+
+wcl::rope Emitter::walk_top(ctx_t ctx, CSTElement node) {
+  wcl::rope_builder builder;
+
+  for (CSTElement child = node.firstChildElement(); !child.empty(); child.nextSiblingElement()) {
+    if (child.isNode()) {
+      builder.append(try_flat(ctx, child));
+      builder.append(newline(ctx));
+    } else {
+      builder.append(walk_token(ctx, child));
+    }
+  }
+
+  // probably shouldn't do this unconditionally
+  builder.drop_last();
```

Comment:
Very concerned about what this is doing, makes me think we got part of this design wrong.

Reply:
The above can't know that a newline/space was just output I think right? I think all of these things, all the walk functions that is, should not output trailing newline/slashes. They never know if they need to do that or not.

Reply:
Line 76 is the newline getting dropped here so we aren't assuming some recursive call added it. My local code doesn't have the if/else anymore so it is always dropping the `newline()`